### PR TITLE
[KLIMADAO/Klimadao] Polygon-digital-carbon for retirement data

### LINKED ISSUE #1938: [subgraph] Migrate to polygon-digital-carbon
Currently site & carbonmark are using the `polygon-bridged-carbon` subgraph for offset and retirement data. This subgraph has been depricated and we should be using `polygon-digital-carbon` https://api.thegraph.com/subgraphs/name/klimadao/polygon-digital-carbon/graphql

### PR DESCRIPTION
**Pull request by psparacino**
## Description

Removes `bridged-carbon` from fetching retirement data for `/retirements` and `/retirements/[index]`.

Eventually `/retirements` will be moved to an endpoint. This is an interim solution for ICR; the ICR retirement data is only available on `digital-carbon`.

**NOTES:**
1) This PR uses several subgraphs deployed to testing (i.e. personal) accounts.  This PR should not be merged until they are deployed under the main klimadao (or maybe now new one for carbonmark?) graph account.
2) I'm using a placeholder ICR logo for now:
Screen Shot 2023-12-19 at 8 47 01 PM

Open question as to what image should be used for the ICR logo

3) **Anything in `site/` can be ignored**

## Related Ticket

Closes #1938 
Also related to #2047 

## Notes For QA

Specific pages, components or journeys that might be affected:
-  /retirements and /retirements/[index]

Relevant preview URLs:
-  Below

Other notes:
-

### CODE REVIEW

File: site/pages/retirements/[beneficiary]/[retirement_index].tsx [hunk 1]
```diff
@@ -4,18 +4,15 @@ import { ParsedUrlQuery } from "querystring";
...
-
+import { getRetirementDetails } from "@klimadao/lib/utils";
 import { getPledgeByAddress } from "components/pages/Pledge/lib/firebase";
 import { Pledge } from "components/pages/Pledge/types";
 import { SingleRetirementPage } from "components/pages/Retirements/SingleRetirement";
 import { getAddressByDomain } from "lib/getAddressByDomain";
 import { getIsDomainInURL } from "lib/getIsDomainInURL";
 import { loadTranslation } from "lib/i18n";
 import { INFURA_ID } from "lib/secrets";
+import { queryKlimaRetireByIndex } from "../../../../carbonmark/lib/retirementDataQueries/retirementDataViaPolygonDigitalCarbon";
```
Comment (0xemc) [inline site/pages/retirements/[beneficiary]/[retirement_index].tsx:15]: This should work right?

```suggestion
import { queryKlimaRetireByIndex } from "lib/retirementDataQueries/retirementDataViaPolygonDigitalCarbon";
```

File: site/pages/retirements/[beneficiary]/index.tsx [hunk 2]
```diff
@@ -11,7 +10,7 @@ import { getIsDomainInURL } from "lib/getIsDomainInURL";
 import { loadTranslation } from "lib/i18n";
 import { INFURA_ID } from "lib/secrets";
 import { GetStaticProps } from "next";
-
+import { queryKlimaRetiresByAddress } from "../../../../carbonmark/lib/retirementDataQueries/retirementDataViaPolygonDigitalCarbon";
```
Comment (0xemc) [inline site/pages/retirements/[beneficiary]/index.tsx:13]: Same as 
```suggestion
import { queryKlimaRetiresByAddress } from "lib/retirementDataQueries/retirementDataViaPolygonDigitalCarbon";
```

File: lib/types/subgraph.ts [hunk 3]
```diff
@@ -1,33 +1,42 @@
 export interface KlimaRetire {
```
Comment (0xemc) [inline lib/types/subgraph.ts:1]: Can this type and `DigitalCarbonCredit` be generated from the API?

File: carbonmark/lib/retirementDataQueries/getProjectInfoViaPolygonDigitalCarbon.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
+import { subgraphs } from "@klimadao/lib/constants";
+
+export const getProjectInfoViaPolygonDigitalCarbon = async (
```
Comment (0xemc) [inline carbonmark/lib/retirementDataQueries/getProjectInfoViaPolygonDigitalCarbon.ts:3]: Hold up.. shouldn't this be handled by the API?

Sorry if you already mentioned this elsewhere.

File: carbonmark/pages/retirements/[beneficiary]/index.tsx [hunk 5]
```diff
@@ -58,7 +59,14 @@ export const getStaticProps: GetStaticProps<Props, Params> = async (ctx) => {
 
       try {
         totalCarbonRetired = klimaRetires
-          .reduce((acc, retirement) => acc + parseFloat(retirement.amount), 0)
+          .reduce((acc, retirement) => {
+            const amount = retirement.retire.credit.project.registry.startsWith(
+              "ICR"
+            )
+              ? retirement.retire.amount
+              : formatUnits(retirement.retire.amount);
```
Comment (0xemc) [inline carbonmark/pages/retirements/[beneficiary]/index.tsx:67]: eww.. lol

File: site/components/pages/Retirements/SingleRetirement/index.tsx [hunk 6]
```diff
@@ -62,44 +60,66 @@ export const SingleRetirementPage: NextPage<SingleRetirementPageProps> = ({
...
+    | null;
 
-  const retiree =
-    retirement.beneficiary ||
-    props.nameserviceDomain ||
-    concatAddress(props.beneficiaryAddress);
+  let tokenData: {
+    key: string;
+    icon: StaticImageData;
+    label: "UBO" | "NBO" | "BCT" | "NCT" | "MCO2" | "TCO2" | "C3T" | "ICR";
```
Comment (0xemc) [inline site/components/pages/Retirements/SingleRetirement/index.tsx:80]: I feel like we should have some defined types already for these possible strings?

File: site/pages/retirements/[beneficiary]/[retirement_index].tsx (hunk 1 above)
Comment (psparacino) [reply site/pages/retirements/[beneficiary]/[retirement_index].tsx]: I changed the above as I was receiving errors on the local test build-all. Committed the suggestion here.

File: site/components/pages/Retirements/SingleRetirement/index.tsx (hunk 6 above)
Comment (psparacino) [reply site/components/pages/Retirements/SingleRetirement/index.tsx:80]: this is in `site/` and can be ignored. I'm going to remove this from the PR

File: carbonmark/pages/retirements/[beneficiary]/index.tsx (hunk 5 above)
Comment (psparacino) [reply carbonmark/pages/retirements/[beneficiary]/index.tsx:67]: Very open to any suggestions on this! There is some refactoring coming down the pipeline that will aim to standardize registry handling by the time it reaches the api or app. Right now the subgraph handles all amounts in wei; it probably makes sense to move this handling into that layer.

File: carbonmark/lib/retirementDataQueries/getProjectInfoViaPolygonDigitalCarbon.ts (hunk 4 above)
Comment (psparacino) [reply carbonmark/lib/retirementDataQueries/getProjectInfoViaPolygonDigitalCarbon.ts:3]: Creating the github issue as discussed

File: lib/types/subgraph.ts (hunk 3 above)
Comment (psparacino) [reply lib/types/subgraph.ts:1]: I will add this to the GH issue. It should be straightforward once these queries are moved to the api.

File: lib/constants/index.ts [hunk 7]
```diff
@@ -276,11 +276,15 @@ export const offsetCompatibility: CompatMap = {
 
 const SUBGRAPH_URL = "https://api.thegraph.com/subgraphs/name/klimadao";
 export const subgraphs = {
-  polygonBridgedCarbon: `${SUBGRAPH_URL}/polygon-bridged-carbon`,
+  // polygonBridgedCarbon: `${SUBGRAPH_URL}/polygon-bridged-carbon`,
+  // @todo temporary before deployed under /klimadao
+  polygonDigitalCarbon: `https://api.thegraph.com/subgraphs/name/psparacino/digital-carbon`,
```
Comment (Atmosfearful) [inline lib/constants/index.ts:281]: Safe to put this on prod until we get the official one? I assume yes just double checking.

**Review by Atmosfearful — COMMENTED** [review]
Just pinging @0xtapi for QA, even though I trust typescript and our schemas, i'm still worried about regressions

These are the pages that touched
- retirement pages `/retirement/[beneficiary_address]/[index]`
- the retirement PDF (click download button on above page)
- retire from portfolio form `/portfolio/[added_address/retire`

**Review by Atmosfearful — APPROVED** [review]
otherwise LGTM after QA thumbs ups

**Comment by 0xtapi** [comment]
> Just pinging @0xtapi for QA, even though I trust typescript and our schemas, i'm still worried about regressions
> 
> These are the pages that touched
> 
> * retirement pages `/retirement/[beneficiary_address]/[index]`
> * the retirement PDF (click download button on above page)
> * retire from portfolio form `/portfolio/[token_address]/retire`

Checked pages, PDF, retired some carbon... found no issues on any of those pages/actions ✅

**Review by 0xtapi — APPROVED** [review]

File: lib/constants/index.ts (hunk 7 above)
Comment (psparacino) [reply lib/constants/index.ts:281]: Just double-checked and I believe so. The only change here from the deployed version is the addition of the hash to the Retire entity. If there are no other blockers we could just update the main deployment with this version and keep an item off of the future to-do list?

**Review by 0xemc — APPROVED** [review]

**Comment by 0xemc** [comment]
@psparacino where are we at with this?

**Comment by psparacino** [comment]
@0xemc The [staging subgraph](https://thegraph.com/hosted-service/subgraph/klimadao/staging-polygon-digital-carbon) is now fully synced so I can swap that one in and use that for the time being until the [main one](https://thegraph.com/hosted-service/subgraph/klimadao/polygon-digital-carbon) is deployed/synced. I don't think there any reason that won't work for the interim in order to get this merged

**Comment by 0xemc** [comment]
> @0xemc The [staging subgraph](https://thegraph.com/hosted-service/subgraph/klimadao/staging-polygon-digital-carbon) is now fully synced so it should be an immediate deploy to main once the PR is merged in the subgraph repo. All the prod subgraphs have been updated in this pr so once the main deployment is ready we can just hit merge.

Cool thanks, can you link the dependent subgraph repo PR here please?

**Comment by psparacino** [comment]
@0xemc Repo: https://github.com/KlimaDAO/klima-subgraph

Once this [PR](https://github.com/KlimaDAO/klima-subgraph/pull/86) is merged as far as I understand the main deployment should be live almost immediately

**Comment by psparacino** [comment]
@0xemc Just fyi, there was a bump in the road and although it's been deployed to main it'll now (unexpectedly) require almost a day to sync again. The staging subgraph is up to date however. I can put that one in for now and then do a quick fix PR tomorrow when the main deployment syncs. Or wait and just merge all at once, whichever works best

File: lib/constants/index.ts [hunk 8]
```diff
@@ -275,12 +275,13 @@ export const offsetCompatibility: CompatMap = {
...
 const SUBGRAPH_URL = "https://api.thegraph.com/subgraphs/name/klimadao";
+const SUBGRAPH_URL_ID = "https://api.thegraph.com/subgraphs/id";
 export const subgraphs = {
-  polygonBridgedCarbon: `${SUBGRAPH_URL}/polygon-bridged-carbon`,
+  polygonDigitalCarbon: `${SUBGRAPH_URL}/polygon-digital-carbon`,
   userCarbon: `${SUBGRAPH_URL}/klimadao-user-carbon`,
   cujoRefiHoldings:
     "https://api.thegraph.com/subgraphs/name/cujowolf/klima-refi-current-holdings",
-  carbonmark: "https://api.thegraph.com/subgraphs/name/najada/marketplace-new", // TODO: ensure when switching drom testnet to mainnet that this is still correct!
+  carbonmark: `${SUBGRAPH_URL_ID}/QmTfM5TvokuyKXCGWL3wTLZ1z4TqRtQ7PdVfQRVU39ovUC`, // TODO: ensure when switching drom testnet to mainnet that this is still correct!
```
Comment (0xemc) [inline lib/constants/index.ts:284]: @psparacino Can you update the comment here?

**Comment by 0xemc** [comment]
@psparacino it's still not clear to me why it was necessary to wait for the subgraph PR since the preview build was passing with the updated subgraph url in `lib/constants.ts`? 

This suggests to me we are not capturing or covering some possible source of bugs in our build pipeline. Can you TLDR this for me?

**Comment by psparacino** [comment]
@0xemc The essential difference is that this PR requires the hash attribute on the retire entity (which the current subgraph deployed to `klimadao/polygon-digital-carbon` does not have). 

Everything builds but the page will 404 if used with the current main subgraph. Just tried both locally with old and new to confirm.

**Pull request merged**